File: c++/93_Restore_IP_Addresses.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
    	vector<string> res;
    	if (s.size() > 12 or s.size() == 0)
    		return res;
    	string subStr;
    	dfs(s, 0, 0, subStr, res);
    	return res;
    }

    void dfs(string s, int idx, int pos, string subStr, vector<string> &res) {
    	if (idx > 3 or pos >= s.size())
    		return;
    	if (idx == 3 and pos >= s.size() - 3) {
    		//subStr += s[pos] + s[pos + 1] + s[pos + 2]; *** this part is wrong, the right is like a computation
    		if (pos == s.size() - 1 and isValid(s, pos, 1))
    			res.push_back(subStr + s[pos]);
    		else if (pos == s.size() - 2 and isValid(s, pos, 2)) {
    			for (int i = pos; i < s.size(); i++)
    				subStr.push_back(s[i]);
    			res.push_back(subStr);
    		}
    		else if (pos == s.size() - 3 and isValid(s, pos, 3)) {
    			for (int i = pos; i < s.size(); i++)
    				subStr.push_back(s[i]);
    			res.push_back(subStr);
    		}
    		return;
    	}
    	if (isValid(s, pos, 1))
    		dfs(s, idx + 1, pos + 1, subStr + s[pos] + '.', res);
    	if (isValid(s, pos, 2))
    		dfs(s, idx + 1, pos + 2, subStr + s[pos] + s[pos + 1] + '.', res);
   		if (isValid(s, pos, 3))
   			dfs(s, idx + 1, pos + 3, subStr + s[pos] + s[pos + 1] + s[pos + 2] + '.', res);
    }
    bool isValid(string s, int pos, int len) {
    	if (len == 1 and pos < s.size())
    		return 1;
    	else if (len == 2 and pos + 1 < s.size() and s[pos] != '0')
    		return 1;
    	else if (len == 3 and pos + 2 < s.size() and s[pos] != '0') {
    		if ((s[pos] - '0') * 100 + (s[pos + 1] - '0') * 10 + s[pos + 2] - '0' < 256)
    			return 1;
    	}
    	return 0;
    }
};


int main() {
	Solution s;
	//vector<string> res = s.restoreIpAddresses("25525511135");
	//vector<string> res = s.restoreIpAddresses("1111");
	//vector<string> res = s.restoreIpAddresses("010010");
	vector<string> res = s.restoreIpAddresses("172162541");
	for (int i = 0; i < res.size(); i++)
		cout << res[i] << endl;
}

File: python/93_Restore_IP_Addresses.py
#!/usr/bin/python
import unittest

'''
For example:
Given "25525511135",

return ["255.255.11.135", "255.255.111.35"]. (Order does not matter)
'''

class Solution(object):
    def restoreIpAddresses(self, s):
        if len(s) > 12 or len(s) == 0:
        	return []
        res = []
        self.dfs(s, 0, 0, "", res)
        return res

    def dfs(self, s, idx, pos, subStr, res): # idx-th comma
    	if idx > 3 or pos >= len(s):
    		return
    	if idx == 3 and int(s[pos:]) < 256:
    		if pos < len(s) - 1 and s[pos] == '0':
    			return
    		res.append(subStr[1:] + "." + s[pos:])
    		return
    	
    	self.dfs(s, idx + 1, pos + 1, subStr + "." + s[pos], res)
    	if s[pos] != '0':
    		self.dfs(s, idx + 1, pos + 2, subStr + "." + s[pos : pos + 2], res)
    	if int(s[pos : pos+3]) < 256 and s[pos] != '0':
    		self.dfs(s, idx + 1, pos + 3, subStr + "." + s[pos : pos + 3], res)



        


class TestSolution(unittest.TestCase):
    def setUp(self):
        self.s = Solution()
        #print self.s.restoreIpAddresses("25525511135")

    def test_method(self):
        """Such as self.assertEqual, self.assertTrue"""
        self.assertEqual(self.s.restoreIpAddresses("25525511135"), ["255.255.11.135", "255.255.111.35"])
        self.assertEqual(self.s.restoreIpAddresses("010010"), ["0.10.0.10","0.100.1.0"])


if __name__ == "__main__":
    unittest.main()